### ISSUE TITLE
Error message is misleading and not helpful when using descriptors

### ISSUE THREAD

**Issue by danpascu** [microsoft/pyright#8181]
Consider this code:

```py

class Base: ...

class BAttribute[T]:
    def __get__(self, instance: Base | None, owner: type[Base] | None = None) -> T: ...
    def __set__(self, instance: Base, value: T) -> None: ...

class OAttribute[T]:
    def __get__(self, instance: object | None, owner: type | None = None) -> T: ...
    def __set__(self, instance: object, value: T) -> None: ...

class Derived(Base):
    battr = BAttribute[int]()
    oattr = OAttribute[int]()

class Other:
    battr = BAttribute[int]()
    oattr = OAttribute[int]()

d = Derived()
o = Other()

d.battr = 1
d.oattr = 1

o.oattr = 1
o.battr = 1           # error: Cannot assign to attribute "battr" for class "Other":  "Literal[1]" is incompatible with "int"
o.battr = int('123')  # error: Cannot assign to attribute "battr" for class "Other":  "int" is incompatible with "int"
```

The comments show the error messages that pyright reports on those lines

As seen in the code, `BAttribute` can only work as a descriptor on a class derived from `Base`, while `OAttribute` can work anywhere. As a result on `Derived` I can use both without error, but on `Other` I can only use `OAttribute` while `BAttribute` will give errors, but the error messages are misleading and confusing ("int" is incompatible with "int").

I'd like to see some error messages that indicate the real problem (that `BAttribute` can only accept instances of Base in its `__get__` and `__set__` methods).

For reference here are the mypy errors that correctly identifies the issue:

```
typing-2.py:32: error: Argument 1 to "__get__" of "BAttribute" has incompatible type "Other"; expected "Base | None"  [arg-type]
typing-2.py:32: error: Argument 2 to "__get__" of "BAttribute" has incompatible type "type[Other]"; expected "type[Base] | None"  [arg-type]
typing-2.py:32: error: Argument 1 to "__set__" of "BAttribute" has incompatible type "Other"; expected "Base"  [arg-type]
typing-2.py:33: error: Argument 1 to "__get__" of "BAttribute" has incompatible type "Other"; expected "Base | None"  [arg-type]
typing-2.py:33: error: Argument 2 to "__get__" of "BAttribute" has incompatible type "type[Other]"; expected "type[Base] | None"  [arg-type]
typing-2.py:33: error: Argument 1 to "__set__" of "BAttribute" has incompatible type "Other"; expected "Base"  [arg-type]
```

**Comment by danpascu** [comment]
Didn't include this in the example to keep it short and I forgot to mention, but obviously this also applies to `__delete__`

**Comment by erictraut** [comment]
Yeah, these error messages are pretty bad. I'll work on improving them.

**Comment by erictraut** [comment]
This is included in pyright 1.1.369.
